pkg/core/resources/apis/mesh/traffic_route_validator.go

```diff
@@ -44,10 +44,99 @@ func (d *TrafficRouteResource) validateConf() (err validators.ValidationError) {
+	if modify.GetPath() != nil {
+		err.Add(d.validateModificationPath(pathBuilder.Field("path"), modify.GetPath(), match))
+	}
+	if modify.GetHost() != nil {
+		err.Add(d.validateModificationHost(pathBuilder.Field("host"), modify.GetHost()))
+	}
+	if modify.GetRequestHeaders() != nil {
+		err.Add(d.validateModificationHeaders(pathBuilder.Field("requestHeaders"), modify.GetRequestHeaders()))
+	}
+	if modify.GetResponseHeaders() != nil {
+		err.Add(d.validateModificationHeaders(pathBuilder.Field("responseHeaders"), modify.GetResponseHeaders()))
+	}
+	return
+}
+
+func (d *TrafficRouteResource) validateModificationPath(
+	pathBuilder validators.PathBuilder,
+	path *mesh_proto.TrafficRoute_Http_Modify_Path,
+	match *mesh_proto.TrafficRoute_Http_Match,
+) (err validators.ValidationError) {
+	switch path.Type.(type) {
+	case *mesh_proto.TrafficRoute_Http_Modify_Path_RewritePrefix:
+		if path.GetRewritePrefix() == "" {
+			err.AddViolationAt(pathBuilder.Field("rewritePrefix"), "cannot be empty")
+		}
+		if path.GetRewritePrefix() != "" && match.GetPath().GetPrefix() == "" {
+			err.AddViolationAt(pathBuilder.Field("rewritePrefix"), "can only be set when .http.match.path.prefix is not empty")
+		}
+	case *mesh_proto.TrafficRoute_Http_Modify_Path_Regex:
+		if path.GetRegex().GetPattern() == "" {
+			err.AddViolationAt(pathBuilder.Field("regex").Field("pattern"), "cannot be empty")
+		}
+		if path.GetRegex().GetSubstitution() == "" {
+			err.AddViolationAt(pathBuilder.Field("regex").Field("substitution"), "cannot be empty")
+		}
+	default:
+		err.AddViolationAt(pathBuilder, `either "rewritePrefix" or "regex" has to be set`)
+	}
+	return
+}
+
+func (d *TrafficRouteResource) validateModificationHost(
+	pathBuilder validators.PathBuilder,
+	host *mesh_proto.TrafficRoute_Http_Modify_Host,
+) (err validators.ValidationError) {
+	switch host.Type.(type) {
+	case *mesh_proto.TrafficRoute_Http_Modify_Host_Value:
+		if host.GetValue() == "" {
+			err.AddViolationAt(pathBuilder.Field("value"), "cannot be empty")
+		}
+	case *mesh_proto.TrafficRoute_Http_Modify_Host_FromPath:
+		if host.GetFromPath().GetPattern() == "" {
+			err.AddViolationAt(pathBuilder.Field("fromPath").Field("pattern"), "cannot be empty")
+		}
+		if host.GetFromPath().GetSubstitution() == "" {
+			err.AddViolationAt(pathBuilder.Field("fromPath").Field("substitution"), "cannot be empty")
+		}
+	default:
+		err.AddViolationAt(pathBuilder, `either "value" or "fromPath" has to be set`)
+	}
+	return
+}
+
+func (d *TrafficRouteResource) validateModificationHeaders(
+	pathBuilder validators.PathBuilder,
+	headers *mesh_proto.TrafficRoute_Http_Modify_Headers,
+) (err validators.ValidationError) {
+	for i, add := range headers.GetAdd() {
+		if add.GetName() == "" {
+			err.AddViolationAt(pathBuilder.Field("add").Index(i).Field("name"), "cannot be empty")
+		}
+		if add.GetValue() == "" {
+			err.AddViolationAt(pathBuilder.Field("add").Index(i).Field("value"), "cannot be empty")
+		}
+	}
```

Comment:
Is it legal to modify the "Host" or ":authority" headers here?

Reply:
it is, but host section takes precedence

Reply:
I think we should enforce the separation. IIUC, envoy will not accept the Host header here because `isModifiableHeader` and `isRemovableHeader` will be false for the "Host" header and any ":"-prefixed pseudo header.

If you think it makes sense to take this to a separate issue and PR, I'm OK with that :)

Reply:
I filed https://github.com/kumahq/kuma/issues/2122